code/datums/components/embedded.dm

```diff
@@ -124,81 +131,105 @@
-/datum/component/embedded/proc/ripOutHuman()
-	var/mob/living/carbon/human/victim = parent
+/datum/component/embedded/proc/ripOutHuman(datum/source, obj/item/I, obj/item/bodypart/limb)
+	if(I != weapon || limb != limb)
+		return
+
+	var/mob/living/carbon/victim = parent
 	var/time_taken = rip_time * weapon.w_class
 
-	victim.visible_message("<span class='warning'>[victim] attempts to remove [weapon] from [victim.p_their()] [L.name].</span>","<span class='notice'>You attempt to remove [weapon] from your [L.name]... (It will take [DisplayTimeText(time_taken)].)</span>")
+	victim.visible_message("<span class='warning'>[victim] attempts to remove [weapon] from [victim.p_their()] [limb.name].</span>","<span class='notice'>You attempt to remove [weapon] from your [limb.name]... (It will take [DisplayTimeText(time_taken)].)</span>")
 	if(do_after(victim, time_taken, target = victim))
-		if(!weapon || !L || weapon.loc != victim || !(weapon in L.embedded_objects))
+		if(!weapon || !limb || weapon.loc != victim || !(weapon in limb.embedded_objects))
 			qdel(src)
+			return
 
 		if(harmful)
 			var/damage = weapon.w_class * remove_pain_mult
-			L.receive_damage(brute=(1-pain_stam_pct) * damage, stamina=pain_stam_pct * damage) //It hurts to rip it out, get surgery you dingus.
+			limb.receive_damage(brute=(1-pain_stam_pct) * damage, stamina=pain_stam_pct * damage) //It hurts to rip it out, get surgery you dingus.
 			victim.emote("scream")
-			victim.visible_message("<span class='notice'>[victim] successfully rips [weapon] out of [victim.p_their()] [L.name]!</span>", "<span class='notice'>You successfully remove [weapon] from your [L.name].</span>")
+			victim.visible_message("<span class='notice'>[victim] successfully rips [weapon] out of [victim.p_their()] [limb.name]!</span>", "<span class='notice'>You successfully remove [weapon] from your [limb.name].</span>")
 		else
-			victim.visible_message("<span class='notice'>[victim] successfully rips [weapon] off of [victim.p_their()] [L.name]!</span>", "<span class='notice'>You successfully remove [weapon] from your [L.name].</span>")
+			victim.visible_message("<span class='notice'>[victim] successfully rips [weapon] off of [victim.p_their()] [limb.name]!</span>", "<span class='notice'>You successfully remove [weapon] from your [limb.name].</span>")
 
 		safeRemoveHuman(TRUE)
 
 
 /// This proc handles the final step and actual removal of an embedded/stuck item from a human, whether or not it was actually removed safely.
 /// Pass TRUE for to_hands if we want it to go to the victim's hands when they pull it out
 /datum/component/embedded/proc/safeRemoveHuman(to_hands)
-	var/mob/living/carbon/human/victim = parent
-	L.embedded_objects -= weapon
+	var/mob/living/carbon/victim = parent
+	limb.embedded_objects -= weapon
+
+	if(!weapon)
+		if(!victim.has_embedded_objects())
+			victim.clear_alert("embeddedobject")
+			SEND_SIGNAL(victim, COMSIG_CLEAR_MOOD_EVENT, "embedded")
+		QDEL_NULL(src)
```

Comment:
Replace every instance of `QDEL_NULL(src)` with `qdel(src)`
There's no need to null `src`.